### PR TITLE
Clean up lexer

### PR DESCRIPTION
**Pull request by wycats**
This commit substantially refactors and cleans up the lexer. It incorporates the recent comment changes.

- Moves the state of the various lexer functions into structs, and turn the state transitions into methods.
- Creates a new `TokensBuilder` convenience struct to represent the logic for in-progress tokens that contain child tokens. This makes "no content yet" more explicit, and consolidates a lot of code that was repeated in ad hoc ways.
- Breaks up the giant lexer file into a handful of smaller files.
- Cleans up the comment code to make the whitespace stripping less "all-or-nothing", so that whitespace can be easily recovered if needed (for example, when using a comment for something other than documentation).

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -18,80 +18,100 @@ members = ["crates/*/"]
...
+    "ptree-support",
+    "term-support",
+    "uuid-support",
+    "rustyline-support",
+    "match",
+    "post",
+    "fetch",
+    "zip-support",
 ]
-extra = ["default", "binaryview", "tree", "clipboard-cli", "trash-support", "start", "bson", "sqlite", "s3", "chart", "xpath", "selector"]
```
Comment (jonathandturner) [inline Cargo.toml:91]: Did you mean to remove the `extra` feature here? This is how we enable all the optional features.

**Comment by LhKipp** [comment]
Glad to see that some work is being done on the lexer.

 One problem I encountered with the lexer when writing parsing code for the parameter / flag list of a definition (`def cmd [param_flag_list] {...}`) was that the lexer does not split baseline tokens on: `,` `?` `:`  and returns the chars `,` `?` `:` as single tokens. This is correct behavior for most of the nushell syntax. However the parameter / flag list is more restricted and it would be really helpful if the lexer would allow some customization points to lex like desired. 

Another annoyance is that the lexer currently does not split words like `--flag(-f)` into 2 tokens (`--flag` and `(-f)`) but returns it as 1 baseline token. Again: It's correct, but in a more restricted area like the parameter flag list, I would wish to have a customization point to make the desired behavior possible.

Do your changes open up the possibility for that? I would be glad if they would :)

**Comment by wycats** [comment]
@LhKipp these are good points. What did you think of the approach I used for comments, which keeps multiple different views of the same data in a single struct, so downstream consumers can decide how to use them?

**Comment by LhKipp** [comment]
@wycats I like this abstraction. It will help make code clearer.

File: crates/nu-parser/src/lex/tokens.rs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        ws: impl Into<Spanned<String>>,
+        comment: impl Into<Spanned<String>>,
+    ) -> LiteComment {
+        LiteComment {
+            leading_ws: Some(ws.into()),
+            rest: comment.into(),
+        }
+    }
+
+    pub fn dedent(&self, excluded_spaces: usize) -> LiteComment {
```
Comment (LhKipp) [inline crates/nu-parser/src/lex/tokens.rs:71]: As a non native english speaker, it is hard for me to figure out what "dedent" means here. Maybe a better name would be: "trim_max"?

**Review by LhKipp — APPROVED** [review]

File: crates/nu-parser/src/lex/tokens.rs (hunk 2 above)
Comment (jonathandturner) [reply crates/nu-parser/src/lex/tokens.rs:71]: or unindent

**Comment by wycats** [comment]
@LhKipp Thanks for the feedback. I think the next step I'll take on the lexer is to try to beef up some speculative lexing for foundational constructs (flags is a good one).

Historically, I tried to avoid putting too many smarts into the first-line lexer, because the external syntax is very flexible (by design, and this is important). And it's also important not to turn flags that start with `--` and don't ultimate match the flag syntax into errors in the lexer (because downstream shapes may still want to allow them). But the approach of adding a few more smarts into the lexer in case downstream syntax wants to use them is good.

fwiw, the main difference between comments and flags is that the lexer already needed all of the relevant comment states to do the basic lexing job, and I wouldn't want to turn the lexer into a really complex state machine just to faciliate work that downstream shapes could do on demand. That said, flags are really foundational and I think it's ok to do some work in the lexer to support them.

**Comment by LhKipp** [comment]
@wycats I personally wouldn't like to have a special Token for flags etc. I think the desire to occasionally lex something as:
`<baseline_token><opening_brace> <baseline_token> <closing_brace>` (Note the absence of a space before the opening brace) as 2 tokens and not as 1 (as it currently gets lexed) is a bit more universal than only flags.

**Comment by wycats** [comment]
@LhKipp This makes a lot of sense to me. I'll tackle braced delimiters next.

**Comment by LhKipp** [comment]
Great :)

**Pull request merged**